Fixed #30091 -- Documented middleware ordering requirements when using CSRF_USE_SESSIONS.

carltongibson opened the pull request:
Error views are wrapped with `@requires_csrf_token`, so SessionMiddleware must appear before any middleware that may raise an exception when using `CSRF_USE_SESSIONS`.

timgraham commented on docs/ref/middleware.txt:
```diff
@@ -428,6 +428,9 @@ Here are some hints about the ordering of various Django middleware classes:
 
    After ``UpdateCacheMiddleware``: Modifies ``Vary`` header.
 
+   It must come before any middleware that may raise an exception if you're
```
To match the style of other phrases, put it above "After" and chop "It must come".

carltongibson commented:
Thanks for the review Tim. I adjusted, added the note to settings, and added a `fixup!` pulling the comment I based my original suggestion on into line. (We can pick/drop the latter as you think.)

timgraham commented on docs/ref/settings.txt:
```diff
@@ -403,6 +403,13 @@ Storing the CSRF token in a cookie (Django's default) is safe, but storing it
 in the session is common practice in other web frameworks and therefore
 sometimes demanded by security auditors.
 
+.. note::
```
Prefer `..admonition: descriptive title` to `..note::` but I think a separate paragraph is fine. Use callout boxes sparingly.

timgraham commented on docs/ref/middleware.txt:
```diff
@@ -450,13 +453,14 @@ Here are some hints about the ordering of various Django middleware classes:
 
-   It must come after ``SessionMiddleware`` if you're using
-   :setting:`CSRF_USE_SESSIONS`.
+   After ``SessionMiddleware`` if you're using :setting:`CSRF_USE_SESSIONS`.
```
Probably not worth changing and requiring retranslation... unless duplicate strings don't require translation twice... @claudep ?

claudep replied:
Sorry, my comment was not posted when I clicked on resolve conversation. I said this this change is OK because identical paragraphs are indeed only translated once.

timgraham commented on docs/ref/settings.txt:
```diff
@@ -403,6 +403,13 @@ Storing the CSRF token in a cookie (Django's default) is safe, but storing it
+
+    Since the default error views use the CSRF token,
+    :class:`~django.contrib.sessions.middleware.SessionMiddleware` must appear
+    in :setting:`MIDDLEWARE` before any middleware that may raise an exception
```
"raise an exception to trigger an error view (such as `PermissionDenied`)" may offer a bit more explanation.

timgraham commented on docs/ref/middleware.txt:
```diff
@@ -426,6 +426,9 @@ Here are some hints about the ordering of various Django middleware classes:
 
 #. :class:`~django.contrib.sessions.middleware.SessionMiddleware`
 
+   Before any middleware that may raise an exception if you're using
```
"to trigger an error view..." (as suggested below)

carltongibson commented:
CI failure is ref #10913 here.
